Level 3

Aghabayova opened the pull request:
функционалюность

Aghabayova committed:
import route component ProtectedRoute

Aghabayova committed:
fix: register

Aghabayova committed:
fix: authorisation

Aghabayova committed:
fix: login and logout

Aghabayova committed:
fix: savedArticelsCount

Aghabayova committed:
fix: generate articles

Aghabayova committed:
fix: saved-news category, NewsApi requests

Aghabayova committed:
fix: navigation saved-news when loggedin

Aghabayova committed:
fix

Aghabayova committed:
fix: errors

Aghabayova committed:
fix: ReadMe

SmilingJey commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [ ] <!--30.27--> Функциональность проекта из брифа реализована полностью:
  - Прелоадер крутится в ожидании выполнения запроса.
  - Все ссылки и кнопки в проекте работают.
  - Все формы валидируются и на стороне клиента.
  **Комментарий: кнопка отображается как активная даже если одно поле невалидно http://prntscr.com/vm3tjn**
  - В модальном окне «Регистрация» клик по кнопке «Зарегистрироваться» при корректно введённых данных отправляет запрос на роут /signup. Если запрос прошёл успешно, появляется сообщение об успешной регистрации.
  - В модальном окне «Вход» клик по кнопке «Войти» при корректно введённых данных отправляет запрос на роут /signin. Если запрос прошёл успешно, модальное окно закрывается.
  - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должна быть кнопка «Авторизоваться»; а если пользователь залогинился, кнопка «Авторизоваться» исчезает. На её месте появляется ссылка «Сохранённые статьи» и кнопка выхода из системы.
  - После успешного сабмита формы поиска появляется блок с результатами. Если ничего не найдено, появляется надпись «Ничего не найдено».
  - В блоке результата отображаются 3 карточки. Нажатие на кнопку «Показать ещё» отображает следующие 3 карточки.
  - Если пользователь закрыл вкладку, а после — вернулся на сайт, данные достаются из локального хранилища при монтировании компонента App.
  **Комментарий: при поиске новостей результаты запроса должны сохраниться в хранилище, и если заново открыть главную страницу должны отобразиться карточки**
  - После авторизации пользователя, в блоке карточки иконка становится активной.
  - При клике на иконку «Сохранить» в блоке карточки выполняется запрос к /articles нашего API.
  - При клике на активную иконку «Сохранить» в блоке карточки выполняется запрос на удаление карточки. После успешного запроса карточка удаляется из раздела «Сохранённые статьи».
  - На странице «Сохранённые статьи» отображаются: имя пользователя, количество сохранённых статей и ключевые слова, по которым статьи были найдены.
  - На странице «Сохранённые статьи» блоки карточек также содержат: ключевое слово, по которому карточка была найдена, и иконка корзины для удаления статьи.
- [x] <!--2.16--> Сборка и запуск проекта выполняются без ошибок.
- [x] <!--2.16--> Создана глобальная стейт-переменная `currentUser`.
- [x] <!--2.16--> Локальное хранилище: работа с `jwt` токеном и результатами поискового запроса выполнена в правильном порядке.
- Регистрация и авторизация:
  - [x] <!--2.16--> Роут `/saved-news` защищён HOC-компонентом `ProtectedRoute.`
  - [x] <!--2.16--> При попытке перейти по роуту /saved-news происходит редирект на / с открытым модальным окном авторизации. 
  **Комментарий: окно не открывается**
  - [x] <!--2.16--> При нажатии на иконку сохранения статьи неавторизованным пользователем открывается модальное окно с предложением зарегистрироваться. 
  - [x] <!--2.16--> После успешной авторизации модальное окно закрывается. А в шапке появляется ссылка на «Сохранённые статьи» и кнопка выхода из системы с именем пользователя.
  - [x] <!--2.16--> Роут / не является защищённым. 
  - [x] <!--2.16--> После успешного вызова обработчика `onSignOut` происходит редирект на главную;
  - [x] <!--2.16--> Корректно используется хук `useHistory`. 
  - [x] <!--2.16--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`.
- Компоненты:
  - [x] <!--2.16--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--2.16--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2.16--> Хуки вызываются в основной функции компонента. 
  - [x] <!--2.16--> Компоненты `Main`, `NavBar` и `SavedNewsHeader` подписаны на контекст `CurrentUserContext`.
  - [x] <!--2.16--> В компонент App внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--2.16--> В корневом компоненте App создана стейт-переменная `currentUser`. Она используется в качестве значения для провайдера контекста.
  - [x] <!--2.16--> В корневом компоненте App описаны обработчики: `onRegister`, `onLogin` и `onSignOut`. Эти обработчики переданы в соответствующие компоненты: `Register.js`, `Login.js` и `NavBar.js`. 
  - [x] <!--2.16--> Компоненты модальных окон содержат только обработчики сабмита формы. Остальные обработчики, например `handleUpdateUser`, описаны в компоненте App. 
- [x] <!--2.16--> Асинхронные запросы к API:
  - запросы можно осуществлять через `Fetch API` или `XMLHttpRequest`; сторонние библиотеки запрещены — такие как axios или jQuery;
  - для работы с API создан отдельный класс;
  - цепочка обработки промисов завершается блоком `catch`;
  - первый обработчик then возрващает `res.json`;
  - манипуляции с DOM-элементами производятся после получения ответа сервера; это касается и сохранения в localStorage;
- [x] <!--2.16--> Именование:
  - camelCase для параметров, методов и переменных;
  **Комментарий: num_word**
  - существительные для имён переменных;
  - существительные во множественном числе для `NodeList`;
  - существительные с заглавной буквы для имён классов;
  - const для неизменяемыx напрямую переменных;
  - имя класса соответствует содержимому;;
  - отсутсвует транслит и неуместные сокращения.
- [x] <!--2.16--> Сторонние библиотеки применяются только для реализации слайдеров, слайдер создаётся из JavaScript.
- [x] <!--2.16--> Дата рассчитывается автоматически: «за 7 дней до сегодняшнего дня»

### Хорошие практики
- [x] <!--1.88--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--1.88--> Запрос к API за информацией о пользователе и массиве карточек выполняется единожды при монтировании. 
- [x] <!--1.88--> Запросы к API описаны внутри компонента App. 
- [x] <!--1.88--> В качестве функции-обработчика не передаются стрелочные функции. 
- [x] <!--1.88--> Обработка ошибок API:
  - [x] <!--1.88--> пользователь получает сообщение в случае ошибки;
  **Комментарий: зачтено с натяжкой, текст ошибки при регистрации и авторизации невсегда соответсвует действительности** 
  - [ ] <!--1.88--> поля формы заблокированы во время отправки запросов. 
  **Комментарий: на время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы**
- [ ] <!--1.88--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: date.setDate(date.getDate() - 7);  во многих местах slice(3) **

### Рекомендации
- [x] <!--1.67--> Используются семантически правильные блоки для компонентов.
- [x] <!--1.67--> Не используются <div> или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--1.67--> Отсутствует «мусор» в коде :
  - по имени коллбэка слушателя понятно его назначение; 
  - нет «магических значений» — все значения записаны в переменные;
  - константы описаны в той области видимости, в которой объявлена функция. Внутри функции значения констант берутся из замыкания. 

## Количество баллов: 66

## Хорошая работа, но баллов пока недостаточно для сдачи. Необходимо поправить как минимум замечания в разделе "Функциональность проекта из брифа реализована полностью"

Aghabayova committed:
fixes

1. Single config file
2. Save news after refresh
3. Protected routes redirect + login popup
4. Real Time input validation
5. Disabled inputs on submit
6. some bugs :)

SmilingJey commented:
# Резюме:
## JavaScript
### Критерии, влияющие на работоспособность
- [x] <!--30.27--> Функциональность проекта из брифа реализована полностью:
  - Прелоадер крутится в ожидании выполнения запроса.
  - Все ссылки и кнопки в проекте работают.
  - Все формы валидируются и на стороне клиента.
  - В модальном окне «Регистрация» клик по кнопке «Зарегистрироваться» при корректно введённых данных отправляет запрос на роут /signup. Если запрос прошёл успешно, появляется сообщение об успешной регистрации.
  - В модальном окне «Вход» клик по кнопке «Войти» при корректно введённых данных отправляет запрос на роут /signin. Если запрос прошёл успешно, модальное окно закрывается.
  - Правильно работают оба состояния шапки: если пользователь не залогинился, в шапке должна быть кнопка «Авторизоваться»; а если пользователь залогинился, кнопка «Авторизоваться» исчезает. На её месте появляется ссылка «Сохранённые статьи» и кнопка выхода из системы.
  - После успешного сабмита формы поиска появляется блок с результатами. Если ничего не найдено, появляется надпись «Ничего не найдено».
  - В блоке результата отображаются 3 карточки. Нажатие на кнопку «Показать ещё» отображает следующие 3 карточки.
  - Если пользователь закрыл вкладку, а после — вернулся на сайт, данные достаются из локального хранилища при монтировании компонента App.
  - После авторизации пользователя, в блоке карточки иконка становится активной.
  - При клике на иконку «Сохранить» в блоке карточки выполняется запрос к /articles нашего API.
  - При клике на активную иконку «Сохранить» в блоке карточки выполняется запрос на удаление карточки. После успешного запроса карточка удаляется из раздела «Сохранённые статьи».
  - На странице «Сохранённые статьи» отображаются: имя пользователя, количество сохранённых статей и ключевые слова, по которым статьи были найдены.
  - На странице «Сохранённые статьи» блоки карточек также содержат: ключевое слово, по которому карточка была найдена, и иконка корзины для удаления статьи.
- [x] <!--2.16--> Сборка и запуск проекта выполняются без ошибок.
- [x] <!--2.16--> Создана глобальная стейт-переменная `currentUser`.
- [x] <!--2.16--> Локальное хранилище: работа с `jwt` токеном и результатами поискового запроса выполнена в правильном порядке.
- Регистрация и авторизация:
  - [x] <!--2.16--> Роут `/saved-news` защищён HOC-компонентом `ProtectedRoute.`
  - [x] <!--2.16--> При попытке перейти по роуту /saved-news происходит редирект на / с открытым модальным окном авторизации. 
  - [x] <!--2.16--> При нажатии на иконку сохранения статьи неавторизованным пользователем открывается модальное окно с предложением зарегистрироваться. 
  - [x] <!--2.16--> После успешной авторизации модальное окно закрывается. А в шапке появляется ссылка на «Сохранённые статьи» и кнопка выхода из системы с именем пользователя.
  - [x] <!--2.16--> Роут / не является защищённым. 
  - [x] <!--2.16--> После успешного вызова обработчика `onSignOut` происходит редирект на главную;
  - [x] <!--2.16--> Корректно используется хук `useHistory`. 
  - [x] <!--2.16--> Корректно используются компоненты `<Switch />`, `<Route />` и `<Redirect />`.
- Компоненты:
  - [x] <!--2.16--> Хуки не используются внутри условных блоков или циклов. 
  - [x] <!--2.16--> При использовании классовых компонентов эффекты описаны внутри методов жизненного цикла компонента. 
  - [x] <!--2.16--> Хуки вызываются в основной функции компонента. 
  - [x] <!--2.16--> Компоненты `Main`, `NavBar` и `SavedNewsHeader` подписаны на контекст `CurrentUserContext`.
  - [x] <!--2.16--> В компонент App внедрён контекст через `CurrentUserContext.Provider`. 
  - [x] <!--2.16--> В корневом компоненте App создана стейт-переменная `currentUser`. Она используется в качестве значения для провайдера контекста.
  - [x] <!--2.16--> В корневом компоненте App описаны обработчики: `onRegister`, `onLogin` и `onSignOut`. Эти обработчики переданы в соответствующие компоненты: `Register.js`, `Login.js` и `NavBar.js`. 
  - [x] <!--2.16--> Компоненты модальных окон содержат только обработчики сабмита формы. Остальные обработчики, например `handleUpdateUser`, описаны в компоненте App. 
- [x] <!--2.16--> Асинхронные запросы к API:
  - запросы можно осуществлять через `Fetch API` или `XMLHttpRequest`; сторонние библиотеки запрещены — такие как axios или jQuery;
  - для работы с API создан отдельный класс;
  - цепочка обработки промисов завершается блоком `catch`;
  - первый обработчик then возрващает `res.json`;
  - манипуляции с DOM-элементами производятся после получения ответа сервера; это касается и сохранения в localStorage;
- [x] <!--2.16--> Именование:
  - camelCase для параметров, методов и переменных;
  - существительные для имён переменных;
  - существительные во множественном числе для `NodeList`;
  - существительные с заглавной буквы для имён классов;
  - const для неизменяемыx напрямую переменных;
  - имя класса соответствует содержимому;;
  - отсутсвует транслит и неуместные сокращения.
- [x] <!--2.16--> Сторонние библиотеки применяются только для реализации слайдеров, слайдер создаётся из JavaScript.
- [x] <!--2.16--> Дата рассчитывается автоматически: «за 7 дней до сегодняшнего дня»

### Хорошие практики
- [x] <!--1.88--> Начальное состояние стейт-переменных содержит корректный тип данных.
- [x] <!--1.88--> Запрос к API за информацией о пользователе и массиве карточек выполняется единожды при монтировании. 
- [x] <!--1.88--> Запросы к API описаны внутри компонента App. 
- [x] <!--1.88--> В качестве функции-обработчика не передаются стрелочные функции. 
- [x] <!--1.88--> Обработка ошибок API:
  - [x] <!--1.88--> пользователь получает сообщение в случае ошибки;
  - [ ] <!--1.88--> поля формы заблокированы во время отправки запросов. 
  **Комментарий: на время выполнения запроса считается хорошей практикой блокировать поля ввода и кнопку отправки формы**
- [ ] <!--1.88--> Фиксированные значения (константы) именуются заглавными буквами и вынесены в отдельный конфиг-файл. 
**Комментарий: date.setDate(date.getDate() - 7)**

### Рекомендации
- [x] <!--1.67--> Используются семантически правильные блоки для компонентов.
- [x] <!--1.67--> Не используются <div> или иные лишние HTML-тэги для компонентов, которые состоят из одноуровневых блоков.
- [x] <!--1.67--> Отсутствует «мусор» в коде :
  - по имени коллбэка слушателя понятно его назначение; 
  - нет «магических значений» — все значения записаны в переменные;
  - константы описаны в той области видимости, в которой объявлена функция. Внутри функции значения констант берутся из замыкания. 

## Количество баллов: 96

## Отлично, почти все замечания исправлены и теперь баллов достаточно для сдачи работы. Поздравляю с завершением дипломного проекта!

### Несколько советов для дальнейшего развития:
- свертайте несколько небольших проектов для себя, для закрепления знаний и что бы можно добавить их в портфолио
- еще немного попрактивокаться с React, например изучить redux
- ходить на собеседования, опыт прохождения собеседований не менее важен знаний теории, и Вы поймете какие знания требуются на рынке и где Вы проседаете
- читайте [habr](https://habr.com/ru/) и слушаете подкаксты:
https://soundcloud.com/everyonecan
https://soundcloud.com/podlodka
https://soundcloud.com/devschacht
https://soundcloud.com/begebot
https://soundcloud.com/web-standards
https://www.youtube.com/channel/UCZeU17nbVfzczAkJVTay9vw
